Syncing an existing Database is not completing

Not yet sure what's causing this, but if you manually sync an existing Database nothing is happening beyond this one line in the logs ...

```
2015-11-23 12:40:33,060 [INFO ] metabase.driver.sync :: Syncing PostgreSQL database 'Example'...
```

So changes to tables/fields/values are not getting picked up.

Just working on reproducing this and I'm seeing this error ...

```
2015-11-23 14:48:00,117 [ERROR] metabase.task.sync-databases :: Error syncing database:  2 #error {
 :cause nth not supported on this type: PersistentArrayMap
 :via
 [{:type java.lang.UnsupportedOperationException
   :message nth not supported on this type: PersistentArrayMap
   :at [clojure.lang.RT nthFrom RT.java 933]}]
 :trace
 [[clojure.lang.RT nthFrom RT.java 933]
  [clojure.lang.RT nth RT.java 883]
  [metabase.driver.sync$mark_inactive_tables_BANG_ invoke sync.clj 50]
  [metabase.driver.sync$_sync_database_BANG_ invoke sync.clj 86]
  [metabase.driver.sync$_sync_database_with_tracking_BANG_ invoke sync.clj 100]
  [clojure.core$partial$fn__4529 invoke core.clj 2499]
  [metabase.driver.generic_sql$sync_in_context$fn__25704 invoke generic_sql.clj 85]
  [metabase.driver.generic_sql.util$_with_jdbc_metadata invoke util.clj 46]
  [metabase.driver.generic_sql$sync_in_context invoke generic_sql.clj 84]
  [metabase.driver$eval9794$fn__9964$G__9753__9973 invoke driver.clj 36]
  [metabase.driver.sync$sync_database_BANG_ invoke sync.clj 111]
  [metabase.driver$sync_database_BANG_ invoke driver.clj 337]
  [metabase.task.sync_databases.SyncDatabases$iter__33615__33619$fn__33620$fn__33626 invoke sync_databases.clj 26]
  [metabase.task.sync_databases.SyncDatabases$iter__33615__33619$fn__33620 invoke sync_databases.clj 24]
  [clojure.lang.LazySeq sval LazySeq.java 40]
  [clojure.lang.LazySeq seq LazySeq.java 49]
  [clojure.lang.RT seq RT.java 507]
  [clojure.core$seq__4128 invoke core.clj 137]
  [clojure.core$dorun invoke core.clj 3012]
  [metabase.task.sync_databases.SyncDatabases execute sync_databases.clj 21]
  [org.quartz.core.JobRunShell run JobRunShell.java 213]
  [org.quartz.simpl.SimpleThreadPool$WorkerThread run SimpleThreadPool.java 557]]}
2015-11-23 14:48:00,118 [ERROR] org.quartz.core.JobRunShell :: Job DEFAULT.metabase.task.sync-databases.job threw an unhandled Exception: 
java.lang.NullPointerException
	at clojure.lang.Numbers.ops(Numbers.java:1013)
	at clojure.lang.Numbers.isPos(Numbers.java:96)
	at clojure.core$dorun.invoke(core.clj:3012)
	at metabase.task.sync_databases.SyncDatabases.execute(sync_databases.clj:21)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557)
2015-11-23 14:48:00,119 [ERROR] org.quartz.core.ErrorLogger :: Job (DEFAULT.metabase.task.sync-databases.job threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.NullPointerException]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:224)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557)
Caused by: java.lang.NullPointerException
	at clojure.lang.Numbers.ops(Numbers.java:1013)
	at clojure.lang.Numbers.isPos(Numbers.java:96)
	at clojure.core$dorun.invoke(core.clj:3012)
	at metabase.task.sync_databases.SyncDatabases.execute(sync_databases.clj:21)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	... 1 more

```

--- a/src/metabase/api/database.clj
+++ b/src/metabase/api/database.clj
@@ -178,7 +178,8 @@
   [id]
   (let-404 [db (Database id)]
     (write-check db)
-    (future (driver/sync-database! db))) ; run sync-tables asynchronously
+    ;; just publish a message and let someone else deal with the logistics
+    (events/publish-event :database-trigger-sync db))
   {:status :ok})
 
 

--- a/src/metabase/driver/sync.clj
+++ b/src/metabase/driver/sync.clj
@@ -47,8 +47,8 @@
 (defn- mark-inactive-tables!
   "Mark any `Tables` that are no longer active as such. These are ones that exist in the DB but didn't come back from `active-tables`."
   [database active-tables existing-table->id]
-  (doseq [[[{table :name, schema :schema, :as table}] table-id] existing-table->id]
-    (when-not (contains? active-tables table)
+  (doseq [[{table :name, schema :schema, :as table} table-id] existing-table->id]
+    (when-not (contains? (set (map :name active-tables)) table)
       (upd Table table-id :active false)
       (log/info (u/format-color 'cyan "Marked table %s.%s%s as inactive." (:name database) (if schema (str schema \.) "") table))
 

--- a/src/metabase/events/sync_database.clj
+++ b/src/metabase/events/sync_database.clj
@@ -10,7 +10,8 @@
 
 (def sync-database-topics
   "The `Set` of event topics which are subscribed to for use in database syncing."
-  #{:database-create})
+  #{:database-create
+    :database-trigger-sync})
 
 (def ^:private sync-database-channel
   "Channel for receiving event notifications we want to subscribe to for database sync events."
@@ -28,7 +29,10 @@
     (when-let [{topic :topic object :item} sync-database-event]
       (when-let [database (db/sel :one Database :id (events/object->model-id topic object))]
         ;; just kick off a sync on another thread
-        (future (driver/sync-database! database))))
+        (future (try
+                  (driver/sync-database! database)
+                  (catch Throwable t
+                    (log/error (format "Error syncing Database: %d" (:id database)) t))))))
     (catch Throwable e
       (log/warn (format "Failed to process sync-database event. %s" (:topic sync-database-event)) e))))
 

--- a/src/metabase/task/sync_databases.clj
+++ b/src/metabase/task/sync_databases.clj
@@ -19,7 +19,7 @@
 (jobs/defjob SyncDatabases
   [ctx]
   (dorun
-   (for [database (db/sel :many Database :is_sample false)] ; skip Sample Dataset DB
+    (for [database (db/sel :many Database :is_sample false)] ; skip Sample Dataset DB
       (try
         ;; NOTE: this happens synchronously for now to avoid excessive load if there are lots of databases
         (driver/sync-database! database)

